File: app/src/main/java/com/elseboot3909/GCRClient/ChangeActivity.java
package com.elseboot3909.GCRClient;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import com.elseboot3909.GCRClient.API.AccountAPI;
import com.elseboot3909.GCRClient.Entities.ServerData;
import com.elseboot3909.GCRClient.Utils.ServerDataManager;
import com.elseboot3909.GCRClient.databinding.ActivityChangeBinding;

import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.scalars.ScalarsConverterFactory;

public class ChangeActivity extends AppCompatActivity {

    ActivityChangeBinding binding;

    FragmentTransaction fragmentTransaction;
    Fragment currentFragment;
    Fragment searchFragment;
    Bundle bundle;
    String mName;
    String id;
    boolean isStarred;

    Integer mProgressBarRequests = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        overridePendingTransition(R.anim.enter_from_right, R.anim.quit_to_left);

        binding = ActivityChangeBinding.inflate(getLayoutInflater());

        bundle = getIntent().getExtras();
        id = bundle.getString("id");
        String subject = bundle.getString("subject");
        binding.subject.setText(subject);

        String status = bundle.getString("status");
        switch (status) {
            case "NEW":
                binding.active.status.setText(R.string.active);
                binding.active.status.setVisibility(View.VISIBLE);
                break;
            case "MERGED":
                binding.merged.status.setText(R.string.merged);
                binding.merged.status.setVisibility(View.VISIBLE);
                break;
            case "ABANDONED":
                binding.abandoned.status.setText(R.string.abandoned);
                binding.abandoned.status.setVisibility(View.VISIBLE);
                break;
        }

        if (bundle.getBoolean("work_in_progress")) {
            binding.wip.status.setText(R.string.wip);
            binding.wip.status.setVisibility(View.VISIBLE);
        }

        fragmentTransaction = getSupportFragmentManager().beginTransaction();
        mName = ChangeInfoFragment.class.getName();
        searchFragment = getSupportFragmentManager().findFragmentByTag(mName);
        if (searchFragment == null) {
            ChangeInfoFragment fragment = new ChangeInfoFragment();
            fragment.setArguments(bundle);
            fragmentTransaction
                    .replace(R.id.change_container, fragment, mName)
                    .addToBackStack(mName);
        } else {
            fragmentTransaction
                    .replace(R.id.change_container, searchFragment, mName);
        }
        fragmentTransaction.commit();

        binding.exit.setOnClickListener(view -> finish());

        isStarred = bundle.getBoolean("starred");
        if (isStarred) {
            binding.star.setImageResource(R.drawable.ic_baseline_star);
        }
        binding.star.setOnClickListener(view -> updateStarredChange());

        binding.bottomNavMain.setOnItemSelectedListener((item) -> {
            currentFragment = getSupportFragmentManager().findFragmentById(R.id.change_container);
            fragmentTransaction = getSupportFragmentManager().beginTransaction();
            switch(item.getItemId()) {
                case R.id.info:
                    if (!(currentFragment instanceof ChangeInfoFragment)) {
                        mName = ChangeInfoFragment.class.getName();
                        searchFragment = getSupportFragmentManager().findFragmentByTag(mName);
                        fragmentTransaction.setCustomAnimations(R.anim.enter_from_left, R.anim.quit_to_right);
                        if (searchFragment == null) {
                            ChangeInfoFragment fragment = new ChangeInfoFragment();
                            fragment.setArguments(bundle);
                            fragmentTransaction
                                    .replace(R.id.change_container, fragment, mName)
                                    .addToBackStack(mName);
                        } else {
                            fragmentTransaction
                                    .replace(R.id.change_container, searchFragment, mName);
                        }
                        fragmentTransaction.commit();
                    }
                    return true;
                case R.id.code:
                    if (!(currentFragment instanceof CodeFragment)) {
                        if (currentFragment instanceof ChangeInfoFragment) {
                            fragmentTransaction.setCustomAnimations(R.anim.enter_from_right, R.anim.quit_to_left);
                        } else {
                            fragmentTransaction.setCustomAnimations(R.anim.enter_from_left, R.anim.quit_to_right);
                        }
                        mName = CodeFragment.class.getName();
                        searchFragment = getSupportFragmentManager().findFragmentByTag(mName);
                        if (searchFragment == null) {
                            fragmentTransaction
                                    .replace(R.id.change_container, new CodeFragment(), mName)
                                    .addToBackStack(mName);
                        } else {
                            fragmentTransaction
                                    .replace(R.id.change_container, searchFragment, mName);
                        }
                        fragmentTransaction.commit();
                    }
                    return true;
                case R.id.vote:
                    if (!(currentFragment instanceof VoteFragment)) {
                        if (currentFragment instanceof ChangeInfoFragment || currentFragment instanceof CodeFragment) {
                            fragmentTransaction.setCustomAnimations(R.anim.enter_from_right, R.anim.quit_to_left);
                        } else {
                            fragmentTransaction.setCustomAnimations(R.anim.enter_from_left, R.anim.quit_to_right);
                        }
                        mName = VoteFragment.class.getName();
                        searchFragment = getSupportFragmentManager().findFragmentByTag(mName);
                        if (searchFragment == null) {
                            fragmentTransaction
                                    .replace(R.id.change_container, new VoteFragment(), mName)
                                    .addToBackStack(mName);
                        } else {
                            fragmentTransaction
                                    .replace(R.id.change_container, searchFragment, mName);
                        }
                        fragmentTransaction.commit();
                    }
                    return true;
                case R.id.log:
                    /* TODO */
                    return false;
                case R.id.comment:
                    /* TODO */
                    return false;
                default:
                    return false;
            }
        });

        setContentView(binding.getRoot());
    }

    private void updateStarredChange() {
        binding.star.setClickable(false);
        progressBarManager(true);

        ServerData serverData = ServerDataManager.serverDataList.get(ServerDataManager.selectedPos);
        OkHttpClient client = ServerDataManager.getAuthenticatorClient(serverData.getUsername(), serverData.getPassword());

        Retrofit retrofit = new Retrofit.Builder()
                .client(client)
                .baseUrl(serverData.getServerName() + serverData.getServerNameEnding())
                .addConverterFactory(ScalarsConverterFactory.create())
                .build();

        AccountAPI accountAPI = retrofit.create(AccountAPI.class);
        Call<String> retrofitRequest;
        if (isStarred) {
            retrofitRequest = accountAPI.removeStarredChange(id);
        } else {
            retrofitRequest = accountAPI.putStarredChange(id);
        }

        retrofitRequest.enqueue(new Callback<String>() {
            @Override
            public void onResponse(@NonNull Call<String> call, @NonNull Response<String> response) {
                progressBarManager(false);
                isStarred = !isStarred;
                if (isStarred) {
                    binding.star.setImageResource(R.drawable.ic_baseline_star);
                } else {
                    binding.star.setImageResource(R.drawable.ic_outline_star);
                }
                binding.star.setClickable(true);
            }

            @Override
            public void onFailure(@NonNull Call<String> call, @NonNull Throwable t) {
                progressBarManager(false);
                binding.star.setClickable(true);
            }
        });

    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.enter_from_left, R.anim.quit_to_right);
    }

    @Override
    public void onBackPressed() {
        currentFragment = getSupportFragmentManager().findFragmentById(R.id.change_container);
        if (currentFragment instanceof ChangeInfoFragment) {
            finish();
        } else {
            fragmentTransaction = getSupportFragmentManager().beginTransaction();
            mName = ChangesListFragment.class.getName();
            searchFragment = getSupportFragmentManager().findFragmentByTag(mName);
            if (searchFragment != null) {
                fragmentTransaction.replace(R.id.change_container, searchFragment, mName);
            } else {
                finish();
            }
        }
    }

    public void progressBarManager(boolean enabled) {
        if (enabled) {
            mProgressBarRequests++;
        } else {
            mProgressBarRequests--;
        }
        if (mProgressBarRequests > 0 && binding.progressBar.getVisibility() != View.VISIBLE) {
            binding.progressBar.setVisibility(View.VISIBLE);
        } else if (mProgressBarRequests <= 0 && binding.progressBar.getVisibility() == View.VISIBLE) {
            binding.progressBar.setVisibility(View.GONE);
        }
    }

}

File: app/src/main/java/com/elseboot3909/GCRClient/UI/Change/ChangeActivity.kt
package com.elseboot3909.GCRClient.UI.Change

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.elseboot3909.GCRClient.API.AccountAPI
import com.elseboot3909.GCRClient.Entities.ChangeInfo
import com.elseboot3909.GCRClient.R
import com.elseboot3909.GCRClient.UI.Main.ChangesListFragment
import com.elseboot3909.GCRClient.Utils.Constants
import com.elseboot3909.GCRClient.Utils.NetManager
import com.elseboot3909.GCRClient.databinding.ActivityChangeBinding
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class ChangeActivity : AppCompatActivity() {

    private lateinit var binding: ActivityChangeBinding

    private var mName = ""
    private val bundle = Bundle()
    private var isStarred = false

    private var mProgressBarRequests = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        overridePendingTransition(R.anim.enter_from_right, R.anim.quit_to_left)
        binding = ActivityChangeBinding.inflate(layoutInflater)

        val changeInfo = intent.extras?.getSerializable("changeInfo") as ChangeInfo

        bundle.putString("id", changeInfo.id)
        binding.subject.text = changeInfo.subject

        when (changeInfo.status) {
            "NEW" -> {
                binding.active.status.setText(R.string.active)
                binding.active.status.visibility = View.VISIBLE
            }
            "MERGED" -> {
                binding.active.status.setText(R.string.active)
                binding.active.status.visibility = View.VISIBLE
            }
            "ABANDONED" -> {
                binding.abandoned.status.setText(R.string.abandoned)
                binding.abandoned.status.visibility = View.VISIBLE
            }
        }

        if (changeInfo.work_in_progress) {
            binding.wip.status.setText(R.string.wip)
            binding.wip.status.visibility = View.VISIBLE
        }

        val fragmentTransaction = supportFragmentManager.beginTransaction()
        mName = ChangeInfoFragment::class.java.name
        supportFragmentManager.findFragmentByTag(mName).let {
            if (it != null) {
                fragmentTransaction
                    .replace(R.id.change_container, it, mName)
            } else {
                val fragment = ChangeInfoFragment()
                fragment.arguments = bundle
                fragmentTransaction
                    .replace(R.id.change_container, fragment, mName)
                    .addToBackStack(mName)
            }
        }

        fragmentTransaction.commit()

        binding.exit.setOnClickListener { finish() }

        isStarred = changeInfo.starred
        if (isStarred) {
            binding.star.setImageResource(R.drawable.ic_baseline_star)
        }

        binding.star.setOnClickListener { updateStarredChange(changeInfo.id) }

        binding.bottomNavMain.setOnItemSelectedListener { item ->
            val currentFragment = supportFragmentManager.findFragmentById(R.id.change_container)
            val fgTransaction = supportFragmentManager.beginTransaction()
            when (item.itemId) {
                R.id.info -> {
                    if (currentFragment !is ChangeInfoFragment) {
                        fgTransaction.setCustomAnimations(
                            R.anim.enter_from_left,
                            R.anim.quit_to_right
                        )
                        mName = ChangeInfoFragment::class.java.name
                        supportFragmentManager.findFragmentByTag(mName).let {
                            if (it != null) {
                                fgTransaction
                                    .replace(R.id.change_container, it, mName)
                            } else {
                                val fragment = ChangeInfoFragment()
                                fragment.arguments = bundle
                                fgTransaction
                                    .replace(R.id.change_container, fragment, mName)
                                    .addToBackStack(mName)
                            }
                        }

                        fgTransaction.commit()
                    }
                    true
                }
                R.id.code -> {
                    if (currentFragment !is CodeFragment) {
                        if (currentFragment is ChangeInfoFragment) {
                            fgTransaction.setCustomAnimations(
                                R.anim.enter_from_right,
                                R.anim.quit_to_left
                            )
                        } else {
                            fgTransaction.setCustomAnimations(
                                R.anim.enter_from_left,
                                R.anim.quit_to_right
                            )
                        }
                        mName = CodeFragment::class.java.name
                        supportFragmentManager.findFragmentByTag(mName).let {
                            if (it != null) {
                                fgTransaction
                                    .replace(R.id.change_container, it, mName)
                            } else {
                                val fragment = CodeFragment()
                                fragment.arguments = bundle
                                fgTransaction
                                    .replace(R.id.change_container, fragment, mName)
                                    .addToBackStack(mName)
                            }
                        }
                        fgTransaction.commit()
                    }
                    true
                }
                R.id.vote -> {
                    if (currentFragment !is VoteFragment) {
                        if (currentFragment is ChangeInfoFragment || currentFragment is CodeFragment) {
                            fgTransaction.setCustomAnimations(
                                R.anim.enter_from_right,
                                R.anim.quit_to_left
                            )
                        } else {
                            fgTransaction.setCustomAnimations(
                                R.anim.enter_from_left,
                                R.anim.quit_to_right
                            )
                        }
                        mName = VoteFragment::class.java.name
                        supportFragmentManager.findFragmentByTag(mName).let {
                            if (it != null) {
                                fgTransaction
                                    .replace(R.id.change_container, it, mName)
                            } else {
                                Log.e(Constants.LOG_TAG, "Vote called")
                                val fragment = VoteFragment()
                                fragment.arguments = bundle
                                fgTransaction
                                    .replace(R.id.change_container, fragment, mName)
                                    .addToBackStack(mName)
                            }
                        }
                        fgTransaction.commit()
                    }
                    true
                }
                R.id.log -> false
                R.id.comment -> false
                else -> false
            }
        }

        setContentView(binding.root)
    }

    private fun updateStarredChange(id: String) {
        binding.star.isClickable = false
        progressBarManager(true)

        val retrofit = NetManager.getRetrofitConfiguration(null, true)

        val accountAPI = retrofit.create(AccountAPI::class.java)
        val request: Call<String> = if (isStarred) accountAPI.removeStarredChange(id) else accountAPI.putStarredChange(id)

        request.enqueue(object : Callback<String> {
            override fun onResponse(call: Call<String>, response: Response<String>) {
                progressBarManager(false)
                isStarred = !isStarred
                if (isStarred) {
                    binding.star.setImageResource(R.drawable.ic_baseline_star)
                } else {
                    binding.star.setImageResource(R.drawable.ic_outline_star)
                }
                binding.star.isClickable = true
            }

            override fun onFailure(call: Call<String>, t: Throwable) {
                progressBarManager(false)
                binding.star.isClickable = true
            }

        })
    }

    override fun finish() {
        super.finish()
        overridePendingTransition(R.anim.enter_from_left, R.anim.quit_to_right)
    }

    override fun onBackPressed() {
        val currentFragment = supportFragmentManager.findFragmentById(R.id.change_container)
        if (currentFragment is ChangeInfoFragment) {
            finish()
        } else {
            val fragmentTransaction = supportFragmentManager.beginTransaction()
            mName = ChangesListFragment::class.java.name
            val searchFragment = supportFragmentManager.findFragmentByTag(mName)
            if (searchFragment != null) {
                fragmentTransaction.replace(R.id.change_container, searchFragment, mName)
            } else {
                finish()
            }
        }
    }

    fun progressBarManager(enabled: Boolean) {
        if (enabled) {
            mProgressBarRequests++
        } else {
            mProgressBarRequests--
        }
        if (mProgressBarRequests > 0 && binding.progressBar.visibility != View.VISIBLE) {
            binding.progressBar.visibility = View.VISIBLE
        } else if (mProgressBarRequests <= 0 && binding.progressBar.visibility == View.VISIBLE) {
            binding.progressBar.visibility = View.GONE
        }
    }

}